File: api/src/main/java/com/kagg886/sylu_eoa/model/SchoolCalender.java
package com.kagg886.sylu_eoa.model;

import lombok.AllArgsConstructor;
import lombok.Data;

import java.time.LocalDate;

/**
 * 学期起止时间和周数
 *
 * @author kagg886
 * @date 2023/9/4 11:20
 **/
@Data
@AllArgsConstructor
public class SchoolCalender {
    private LocalDate start;
    private LocalDate end;
    private Term currentTerm;
}

File: eoa-lib/network-core/src/commonMain/kotlin/top/kagg886/sylu_eoa/api/v2/bean/SchoolCalender.kt
package top.kagg886.sylu_eoa.api.v2.bean

import kotlinx.datetime.*
import kotlinx.serialization.Serializable
import kotlinx.datetime.Clock
import kotlin.time.ExperimentalTime

@Serializable
data class SchoolCalender(
    val start: LocalDate,
    val end: LocalDate
) {

    @OptIn(ExperimentalTime::class)
    fun currentWeek(
        now: LocalDate = Clock.System.now().toLocalDateTime(TimeZone.currentSystemDefault()).date
    ): Int {
        if (now < start) {
            return -1
        }

        if (now > end) {
            return -2
        }

        var start = start
        var i = 0
        while (start <= now) {
            i++
            start = start.plus(7, DateTimeUnit.DAY)
        }
        return i
    }

    fun count(): Int = (start.until(end, DateTimeUnit.WEEK) + 1).toInt()
}
